File: app/src/main/java/youtubedownloadhelper/download/DownLoadActivity.java
package youtubedownloadhelper.download;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;


import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;

import youtubedownloadhelper.R;
import youtubedownloadhelper.object.Video;
import youtubedownloadhelper.object.Youtube;
import youtubedownloadhelper.Preferences.SharePerferenceHelper;
import youtubedownloadhelper.download.adapter.DownloadAdapter;
import youtubedownloadhelper.mp3.Mp3Helper;
import youtubedownloadhelper.utils.AndroidUtils;
import youtubedownloadhelper.youtube.YotubeItag;
import youtubedownloadhelper.youtube.YoutubePaserTask;

/**
 * Created by andy on 2015/3/8.
 */
public class DownLoadActivity extends Activity implements AdapterView.OnItemClickListener {
    private final static String TAG = "DownLoadActivity";
    public final static String BUNDLE_KEY_YOUTUBE_ID = "BUNDLE_KEY_YOUTUBE_ID";
    public final static int NOTFTY_ADAPTER = 0;
    public final static int DOWNLOAD_FINISH = 1;
    private TextView tv_title;
    private ImageView iv_title;
    private Context context;
    private Youtube curYoutube;
    private ListView listView;
    private DownloadAdapter downloadAdapter;
    private LinearLayout download_page;
    private RelativeLayout loading_Page;

    private String curVideoId;
    private ImageButton reload;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Mp3Helper.getInstance().init(this);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.activity_download);
        context = this;
        tv_title = (TextView) findViewById(R.id.textView);
        iv_title = (ImageView) findViewById(R.id.imageView);
        reload = (ImageButton) findViewById(R.id.reload);
        listView = (ListView) findViewById(R.id.listview);
        download_page = (LinearLayout) findViewById(R.id.download_page);
        loading_Page = (RelativeLayout) findViewById(R.id.loading_Page);
        findViewById(R.id.button4).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                DownLoadActivity.this.finish();
            }
        });
        reload.setEnabled(false);
        reload.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View v) {
                if(!TextUtils.isEmpty(curVideoId)) {
                    reload.setEnabled(false);
                    youtubeParser(curVideoId);
                }
            }
        });
        downloadAdapter = new DownloadAdapter(this);
        listView.setAdapter(downloadAdapter);
        listView.setOnItemClickListener(this);
        Intent intent = getIntent();
        String action = intent.getAction();
        String type = intent.getType();
        Bundle bd = intent.getExtras();
        if (type != null && action != null && type.equals("text/plain") && Intent.ACTION_SEND.equals(action)) {
            curVideoId = intent.getStringExtra(Intent.EXTRA_TEXT);
            youtubeParser(curVideoId);
        } else if (bd != null) {
            if (bd.containsKey(BUNDLE_KEY_YOUTUBE_ID)) {
                curVideoId = intent.getStringExtra(Intent.EXTRA_TEXT);
                youtubeParser(curVideoId);
            }
        }
    }

    public void youtubeParser(String sharedText) {
        if (!TextUtils.isEmpty(sharedText)) {
            download_page.setVisibility(View.GONE);
            loading_Page.setVisibility(View.VISIBLE);

            new YoutubePaserTask(this, new YoutubePaserTask.CallBack() {
                @Override
                public void success(Youtube youtube) {
                    download_page.setVisibility(View.VISIBLE);
                    loading_Page.setVisibility(View.GONE);
                    curYoutube = youtube;
                    tv_title.setText(curYoutube.getTitle());
                    displayImageUrl(iv_title, curYoutube.getImgeUrl());
                    showDownloadList(curYoutube.getVideoList());
                    reload.setEnabled(true);
                }

                @Override
                public void onfail(String Message) {
                    Toast.makeText(DownLoadActivity.this, Message, Toast.LENGTH_SHORT).show();
                    finish();
                }
            }).execute(sharedText);


        }
    }


    private void displayImageUrl(final ImageView iv, final String imageUrl) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    URL url = new URL(imageUrl);
                    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                    connection.setDoInput(true);
                    connection.connect();
                    InputStream input = connection.getInputStream();
                    final Bitmap bitmap = BitmapFactory.decodeStream(input);
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            iv.setImageBitmap(bitmap);
                        }
                    });
                } catch (IOException e) {
                    e.printStackTrace();
                    return;
                }
            }
        }).start();

    }

    private void showDownloadList(ArrayList<Video> list) {
        if (list == null || list.size() == 0) {
            Toast.makeText(DownLoadActivity.this, "找不到影片", Toast.LENGTH_SHORT).show();
            finish();
            return;
        }
        downloadAdapter.setList(list);
        downloadAdapter.notifyDataSetChanged();
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        Log.d(TAG, "onItemClick " + position);
        if (!downloadAdapter.isDownloading(position)) {
            Video video = downloadAdapter.getItem(position);
            if (video.isDownlaod()) {
                AndroidUtils.playVideo(this, video.getLocalFilePath());
            } else {
                downloadVideo(position, video);
            }
        }
    }

    @Override
    public void onStart() {
        super.onStart();
    }

    @Override
    public void onStop() {
        super.onStop();

    }


    private Handler handler = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(final Message msg) {
            super.handleMessage(msg);
            switch (msg.what){
                case NOTFTY_ADAPTER:
                    if(downloadAdapter != null){
                        downloadAdapter.notifyDataSetChanged();
                    }
                    break;
                case DOWNLOAD_FINISH:
                    if(downloadAdapter != null){
                         Video video = downloadAdapter.getItem(msg.arg1);
                        downloadAdapter.removeDownloadCache(msg.arg1);
                        downloadAdapter.notifyDataSetChanged();


                    }

                    break;
            }
        }
    };

    public void downloadVideo(final int position, final Video video) {
        String fileName = curYoutube.getTitle().trim();
        fileName = fileName.length()>15?fileName.substring(0,15):fileName;
        final DownloadDialog downloadDialog = new DownloadDialog(context, fileName+"_"+YotubeItag.getVideoDescribe(video.getItag())
                                                        + "." + YotubeItag.getVideoType(video.getItag()));
        downloadDialog.setOnFileSelectListener(new DownloadDialog.OnFileSelectListener() {
            @Override
            public void onFileSelected(String filePah) {
                SharePerferenceHelper.getInstance(context).setString("path", filePah);
                DownloadTask task = new DownloadTask(DownLoadActivity.this, handler, position);
                downloadAdapter.putDownloadCache(position, task);
                task.execute(video, filePah, downloadDialog.getCurrentFileName());
            }
        });
        downloadDialog.create().show();
    }


}

File: app/src/main/java/youtubedownloadhelper/download/DownLoadActivity.kt
package youtubedownloadhelper.download

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.view.Window
import android.widget.AdapterView
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ListView
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast


import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.ArrayList

import youtubedownloadhelper.R
import youtubedownloadhelper.`object`.Video
import youtubedownloadhelper.`object`.Youtube
import youtubedownloadhelper.download.adapter.DownloadAdapter
import youtubedownloadhelper.utils.AndroidUtils
import youtubedownloadhelper.youtube.YotubeItag
import youtubedownloadhelper.youtube.YoutubePaserTask

/**
 * Created by andy on 2015/3/8.
 */
class DownLoadActivity : Activity(), AdapterView.OnItemClickListener {
	private var tv_title: TextView? = null
	private var iv_title: ImageView? = null
	private var context: Context? = null
	private var curYoutube: Youtube? = null
	private var listView: ListView? = null
	private var downloadAdapter: DownloadAdapter? = null
	private var download_page: LinearLayout? = null
	private var loading_Page: RelativeLayout? = null

	private var curVideoId: String? = null
	private var reload: ImageButton? = null


	private val handler = object : Handler(Looper.getMainLooper()) {
		override fun handleMessage(msg: Message) {
			super.handleMessage(msg)
			when (msg.what) {
				NOTFTY_ADAPTER -> if (downloadAdapter != null) {
					downloadAdapter!!.notifyDataSetChanged()
				}
				DOWNLOAD_FINISH -> if (downloadAdapter != null) {
					val video = downloadAdapter!!.getItem(msg.arg1)
					downloadAdapter!!.removeDownloadCache(msg.arg1)
					downloadAdapter!!.notifyDataSetChanged()


				}
			}
		}
	}


	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		requestWindowFeature(Window.FEATURE_NO_TITLE)
		setContentView(R.layout.activity_download)
		context = this
		tv_title = findViewById<View>(R.id.textView) as TextView
		iv_title = findViewById<View>(R.id.imageView) as ImageView
		reload = findViewById<View>(R.id.reload) as ImageButton
		listView = findViewById<View>(R.id.listview) as ListView
		download_page = findViewById<View>(R.id.download_page) as LinearLayout
		loading_Page = findViewById<View>(R.id.loading_Page) as RelativeLayout
		findViewById<View>(R.id.button4).setOnClickListener { this@DownLoadActivity.finish() }
		reload!!.isEnabled = false
		reload!!.setOnClickListener {
			if (!TextUtils.isEmpty(curVideoId)) {
				reload!!.isEnabled = false
				youtubeParser(curVideoId)
			}
		}
		downloadAdapter = DownloadAdapter(this)
		listView!!.adapter = downloadAdapter
		listView!!.onItemClickListener = this
		val intent = intent
		val action = intent.action
		val type = intent.type
		val bd = intent.extras
		if (type != null && action != null && type == "text/plain" && Intent.ACTION_SEND == action) {
			curVideoId = intent.getStringExtra(Intent.EXTRA_TEXT)
			youtubeParser(curVideoId)
		} else if (bd != null) {
			if (bd.containsKey(BUNDLE_KEY_YOUTUBE_ID)) {
				curVideoId = intent.getStringExtra(Intent.EXTRA_TEXT)
				youtubeParser(curVideoId)
			}
		}
	}

	fun youtubeParser(sharedText: String?) {
		if (!TextUtils.isEmpty(sharedText)) {
			download_page!!.visibility = View.GONE
			loading_Page!!.visibility = View.VISIBLE

			YoutubePaserTask(this, object : YoutubePaserTask.CallBack {
				override fun success(youtube: Youtube) {
					download_page!!.visibility = View.VISIBLE
					loading_Page!!.visibility = View.GONE
					curYoutube = youtube
					tv_title!!.text = curYoutube!!.title
					displayImageUrl(iv_title, curYoutube!!.imgeUrl)
					showDownloadList(curYoutube!!.videoList)
					reload!!.isEnabled = true
				}

				override fun onfail(Message: String) {
					Toast.makeText(this@DownLoadActivity, Message, Toast.LENGTH_SHORT).show()
					finish()
				}
			}).execute(sharedText)


		}
	}


	private fun displayImageUrl(iv: ImageView?, imageUrl: String?) {
		Thread(Runnable {
			try {
				val url = URL(imageUrl)
				val connection = url.openConnection() as HttpURLConnection
				connection.doInput = true
				connection.connect()
				val input = connection.inputStream
				val bitmap = BitmapFactory.decodeStream(input)
				runOnUiThread { iv!!.setImageBitmap(bitmap) }
			} catch (e: IOException) {
				e.printStackTrace()
				return@Runnable
			}
		}).start()

	}

	private fun showDownloadList(list: ArrayList<Video>?) {
		if (list == null || list.size == 0) {
			Toast.makeText(this@DownLoadActivity, "找不到影片", Toast.LENGTH_SHORT).show()
			finish()
			return
		}
		downloadAdapter!!.setList(list)
		downloadAdapter!!.notifyDataSetChanged()
	}

	override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
		Log.d(TAG, "onItemClick $position")
		if (!downloadAdapter!!.isDownloading(position)) {
			val video = downloadAdapter!!.getItem(position)
			if (video!!.isDownlaod) {
				AndroidUtils.playVideo(this, video.localFilePath!!)
			} else {
				downloadVideo(position, video)
			}
		}
	}

	public override fun onStart() {
		super.onStart()
	}

	public override fun onStop() {
		super.onStop()

	}

	fun downloadVideo(position: Int, video: Video) {
		var fileName = curYoutube!!.title!!.trim { it <= ' ' }
		fileName = if (fileName.length > 15) fileName.substring(0, 15) else fileName
		fileName = (fileName + "_" + YotubeItag.getVideoDescribe(video.itag)
				+ "." + YotubeItag.getVideoType(video.itag))
		val task = DownloadTask(this@DownLoadActivity, handler, position)
		downloadAdapter!!.putDownloadCache(position, task)
		task.execute(video, Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM).absolutePath, fileName)
	}

	companion object {
		private val TAG = "DownLoadActivity"
		val BUNDLE_KEY_YOUTUBE_ID = "BUNDLE_KEY_YOUTUBE_ID"
		val NOTFTY_ADAPTER = 0
		val DOWNLOAD_FINISH = 1
	}


}
